### ISSUE TITLE
[BUG] Expert Parallelism deadlock with Gemma4 (EP=4, DP≥16)

### ISSUE THREAD

**Issue by Aaditya-Singh** [NVIDIA/Megatron-LM#5749]
## Description

When fine-tuning `Gemma4-26B-A4B-it` model with Expert Parallelism (EP=4) on multi-node clusters (64–128 GPUs), all ranks deadlock after 92–852 training steps on a large-scale dataset with multi-image samples. The hang occurs in the NCCL allreduce/reduce-scatter kernel during gradient synchronization, as the EP alltoall collective for MoE token dispatch conflicts with DP gradient reduce-scatter on the same NCCL stream.

The deadlock seems to be caused by the alltoall token dispatcher. An identical configuration with TP=4/EP=1 (eliminating the alltoall collective) completes 1000 steps without hanging on the exact same setup, but suffers from slow step times (7.54s/step v/s 3.44s/step) so EP≥4 is desirable.

## Root Cause Analysis

Progressive patching across 4 runs identified the failure site:

| Run | Patches applied | Hung at step | Blocking site |
| -- | -- | -- | -- |
| v0 | `logical_and_across_model_parallel_group` bypassed | 534 | first CPU sync after stuck GPU |
| v1 | * `reduce_max_stat` bypassed | 720 | next CPU sync |
| v2 | * `check_for_nan_in_grad` disabled | 852 | `finalize_model_grads` (grad allreduce) |
| v3 | all Python sync points removed | 584 | GPU completely stuck in prior allreduce |

The non-deterministic step (534/584/719/852) confirms a timing-dependent NCCL deadlock, not data or iteration-count dependent. py-spy captures show all 128 ranks blocked symmetrically in `logical_and_across_model_parallel_group` (the first Python-level CPU sync that discovers the stuck GPU).

Another experiment on the same cluster, dataset, world_size (128 GPUs), and dependency stack but with `--expert_model_parallel_size 1 --tensor_model_parallel_size 4` (no EP alltoall) completes 1000/1000 steps without hanging. This confirms the source of deadlock to be alltoall collective used for EP.

### Mechanism

The `alltoall` token dispatcher dispatches tokens to EP-remote experts via `torch.distributed.all_to_all`. With EP=4 and `use_distributed_optimizer=True`, the DP gradient reduce-scatter uses `model_parallel_group` communicators that partially overlap with EP communicators. Over hundreds of steps, GPU scheduling jitter causes ranks to enter their respective reduce-scatter collectives out of sync. When one rank starts a new bucket's collective while another is still in alltoall on the same NCCL stream → deadlock.

### Environment

* megatron-core: 0.17.1 (also tested 0.18.1 — same deadlock, hangs earlier)
* Training framework: [ms-swift](<https://github.com/modelscope/ms-swift>) 4.3.1 + [mcore-bridge](<https://github.com/modelscope/mcore-bridge>) 1.5.2 (thin integration layer that calls megatron-core's distributed training APIs)
* PyTorch: 2.8.0a0 (NGC 25.05-py3)
* NCCL: 2.24 (NGC 25.05 bundled); also tested 2.21.5 (NGC 24.10) — same result
* Transformer Engine: 2.3.0 (NGC 25.05 bundled)
* GPU: H200 141GB SXM × 128 (16 nodes × 8 GPUs, EFA 3.2 Tbps)
* Also reproduced on: H200 × 64 (8 nodes × 8 GPUs)

### Model

* Gemma4-26B-A4B-it: 30 layers, 128 experts, 25.8B parameters, hybrid sliding-window + global attention

### Parallelism Configuration

```
TP=1, EP=4, PP=1, DP=32
micro_batch_size=1, global_batch_size=128 (grad_accum=4)
use_distributed_optimizer=True
moe_token_dispatcher_type=alltoall
```

### Minimum Command to Reproduce

```bash
NPROC_PER_NODE=8 NNODES=16 MASTER_ADDR=<master_ip> \
megatron sft \
    --model <path_to_gemma4_26b> \
    --dataset <any_vl_sft_dataset> \
    --tensor_model_parallel_size 1 \
    --expert_model_parallel_size 4 \
    --pipeline_model_parallel_size 1 \
    --micro_batch_size 1 \
    --global_batch_size 128 \
    --attention_backend auto \
    --recompute_granularity full \
    --recompute_method uniform \
    --recompute_num_layers 1 \
    --finetune true \
    --max_length 20480 \
    --train_iters 1000
```

`megatron sft` is the [ms-swift](<https://github.com/modelscope/ms-swift>) CLI for Megatron-backend SFT. The key megatron-core settings are `expert_model_parallel_size=4`, `use_distributed_optimizer=True` (default), and `moe_token_dispatcher_type=alltoall` (default). Any training harness that exercises these codepaths with DP≥16 should reproduce. Training starts normally (3.44 s/step with fused attention) but hangs after 92–852 steps depending on version and timing.

### What Works

* `--expert_model_parallel_size 1 --tensor_model_parallel_size 4 --sequence_parallel true` → **completes 1000 steps** (7.54 s/step, no alltoall) but suffers from slow step times (v/s 3.44s/step for EP=4/TP=1).

### What Doesn't Work

* EP=4 with `alltoall` on 128 GPUs (16 nodes): hangs at step 92–852 (5 separate runs)
* EP=4 with `alltoall` on 64 GPUs (8 nodes): hangs at step 92–139 (2 separate runs)
* megatron-core 0.17.1: hangs at step 852 (3/3 runs, same step each time with same version)
* megatron-core 0.18.1: hangs at step 257 (bucket draining changes timing, triggers race earlier)
* `--overlap_grad_reduce false`: still hangs at step 852
* `--moe_router_num_groups 4 --moe_router_group_topk 1` (balanced routing): hangs at step 103 (worse)

### What Was Ruled Out

| Hypothesis | How tested | Result |
| -- | -- | -- |
| NCCL version | Tested 2.21.5 and 2.24 | Both deadlock |
| Data/seed dependent | 3 different seeds | Same outcome |
| Scale-specific (128 GPU only) | Tested at 64 GPUs (8 nodes) | Also deadlocks |
| `overlap_grad_reduce` timing | `--overlap_grad_reduce false` | Still hangs |
| Token routing asymmetry | Balanced routing (`moe_router_num_groups=4`) | Hung faster |
| Training framework Python sync points | Patched out all 5 blocking `.item()` syncs in training loop | Shifts step, doesn't fix |
| megatron-core bucket management | 0.18.1 `previous_grad_reduce_bucket_group` draining | Hung earlier |
| Distributed optimizer | Can't disable (OOM: 206 GiB Adam states) | Mandatory for 26B |
| EP group size reduction | EP=1 OOMs, EP=2 marginal OOM | Can't reduce |

## Related Issues

* NVIDIA/Megatron-LM#5100 PP deadlock with `variable_seq_lengths=True` and `batch_p2p_comm=True`. Same symptom class (collective ordering conflict causing NCCL deadlock at scale) but different codepath (P2P shape exchange v/s EP alltoall + DP reduce-scatter).

**Comment by cspades** [comment]
Interesting, I wonder if M-FSDP suffers from the same thing, I've seen similar hangs before. Last I checked, we have an independent communicator for EP though.

Even DDP's reduce-scatter too apparently?

**Comment by cspades** [comment]
@Autumn1998 to put this on your radar

**Comment by cspades** [comment]
> moe_token_dispatcher_type=alltoall (default)

@Aaditya-Singh Just curious have you tried the other dispatch backends, e.g. DeepEP, HybridEP with flex dispatcher?

cc @YangFei1990

**Comment by Aaditya-Singh** [comment]
Hi @cspades thanks for following-up, no I haven't given the other dispatch backends a shot yet.

**Comment by cspades** [comment]
See if that unblocks you, I have NCCL people also looking at this since there are bugs tracked on M-FSDP with the same fundamental issues basically. Tell me how it goes!

**Comment by Aaditya-Singh** [comment]
**Update: Tested all available dispatcher backends + `num_distributed_optimizer_instances` (NDOI) on mcore 0.18.0 but all of them deadlock**

Following up on the recommendation to try other dispatch backends, I built a custom image with NCCL 2.30.7 + PyTorch 2.13 (source) + DeepEP 2.1.0 to enable `moe_token_dispatcher_type=flex` with both the `deepep` and `hybridep` backends. Also tested `num_distributed_optimizer_instances=4` as one hypothesis was bucket participation asymmetry in the distributed optimizer.

| Dispatcher | Optimizer | Deadlock Step | Step Time |
|---|---|---|---|
| alltoall (original report) | NDOI=1 | 852 | 3.44 s/step |
| flex+deepep | NDOI=1 | 852 | 3.29 s/step |
| flex+hybridep | NDOI=1 | 852 | 3.29 s/step |
| flex+deepep | NDOI=4 | 600 | 3.42 s/step |
| TP=4/EP=1 (control) | NDOI=1 | None (1000 OK) | 7.54 s/step |

Setup: 128× H200 (16× p5e.48xlarge), Gemma4-26B-A4B EP=4/DP=32.

<details>
<summary>Dependency stack (custom image, built from source)</summary>

```
Base: nvidia/cuda:12.9.0-devel-ubuntu24.04
├── NCCL 2.30.7-1 (built from source, /usr/local/nccl)
├── PyTorch 2.13.0a0 (built from source, USE_SYSTEM_NCCL=1)
├── torchvision 0.22.0 (built from source)
├── DeepEP 2.1.0 (built from source, SM90)
├── Transformer Engine 2.17.0 (built from source, SM90)
├── megatron-core 0.18.0 (pip)
├── mcore-bridge 1.5.2 (pip)
├── ms-swift 4.3.1
├── deepspeed 0.14.4
└── transformers 5.9.0
```

LD_LIBRARY_PATH has `/usr/local/nccl/lib` first (system NCCL 2.26.5 from cudnn apt must NOT be loaded).
</details>

Key observations:
- Deadlock step is deterministic at 852 for any NDOI=1 config, regardless of which dispatcher backend is used. This rules out the token dispatch collectives as the cause.
- `num_distributed_optimizer_instances=4` successfully creates the hierarchical `inter_dist_opt` groups (no assertion errors, training starts fine) but deadlocks 252 steps *earlier*. The additional communicator groups seem to worsen the ordering problem rather than fix it.
- Hang signature is identical in all cases: 100% SM util (NCCL spin-wait), VRAM climbing as grad buffers accumulate without release.

**Comment by YangFei1990** [comment]
@Aaditya-Singh several questions that might help debug
- Is there anything special about the step 852? Does it near any checkpoint saving?
- Do you observe anything in log during the hang?
- I just realized you are running on AWS p5e, are you running with EFA? I wonder how could you run DeepEP/HybridEP on EFA, I think they have strong inifiniteband dependencies

In terms of debugging, you could also try following
- Run with `CUDA_LAUNCH_BLOCKING=1` to see if the hang could be reproduced
- If so above, try to find the exactly hanging point of the different processes and check if there is a mismatch in collective pattern, either through gdb or py-spy

If you can reproduce it with pure megatron, I can also test it on my end.

**Comment by Aaditya-Singh** [comment]
@YangFei1990 Thanks for looking into this. Answers to your questions:

**1. Step 852, anything special?**

There's no checkpointing (`--save-interval 999999`) or eval boundary. Step 852 is timing-dependent: it's consistent at 852 for NDOI=1 across all three dispatcher backends, but shifts (534/584/720) when we patch out Python-level sync points which confirms that it's a race condition whose manifestation step depends on per-step overhead.

**2. Logs during hang?**

- All GPUs at 100% SM (NCCL spin-wait), 0% memory BW
- VRAM climbs from ~81 GiB → 101–122 GiB (grad buffers accumulating without release)
- No Python exception, no NCCL error
- py-spy on all 128 ranks shows symmetric block in `all_reduce` → `.item()` at `logical_and_across_model_parallel_group` (first Python sync that discovers GPU is stuck)
- `AccumulateGrad` stream mismatch warning fires every step: _"The AccumulateGrad node's stream does not match the stream of the node that produced the incoming gradient"_ from `megatron/core/pipeline_parallel/schedules.py:211`

**3. EFA + DeepEP/HybridEP?**

- Yes I'm using AWS EFA (3.2 Tbps/node, p5e.48xlarge) for faster inter-node comms. Both DeepEP and HybridEP seem to use NCCL under the hood rather than InfiniBand verbs, but not sure about their internal dependency structure.
- For testing, I built NCCL 2.30.7 from source with [aws-ofi-nccl](https://github.com/aws/aws-ofi-nccl) and both dispatchers completed 852 steps at 3.29 s/step before deadlocking. The same bug also reproduces on NCCL 2.21.5 and 2.24.

**4. `CUDA_LAUNCH_BLOCKING=1`**

Ran with `CUDA_LAUNCH_BLOCKING=1` + `NCCL_DEBUG=INFO` on 8 nodes (64 GPUs), EP=4, alltoall. Rank 0 crashed at step ~852 (same step as the non-CLB hang). The failure mode changes from an infinite hang to a crash — `ProcessGroupNCCL` detects a broken pipe instead of spin-waiting forever, but the underlying collective ordering failure is the same. CLB just lets the NCCL watchdog fire rather than spinning indefinitely.

**5. Pure megatron repro (no ms-swift)**

Smoke-tested this on a single node with a reduced model (2 layers, hidden=1024) and it runs correctly. The full Gemma4 architecture (30 layers, hidden=2816, 128 experts at EP=4) needs multi-node since it OOMs on a single 8-GPU node. You'll need `--no-gradient-accumulation-fusion` if APEX is not installed:

```bash
# From Megatron-LM repo root. Needs 64+ GPUs (8+ nodes × 8).
torchrun --nproc_per_node=8 --nnodes=8 --node_rank=$RANK \
  --master_addr=$MASTER --master_port=29500 \
  pretrain_gpt.py \
  --mock-data \
  --num-layers 30 \
  --hidden-size 2816 \
  --ffn-hidden-size 11264 \
  --num-attention-heads 32 \
  --group-query-attention \
  --num-query-groups 8 \
  --seq-length 8192 \
  --max-position-embeddings 8192 \
  --position-embedding-type rope \
  --rotary-base 1000000 \
  --normalization RMSNorm \
  --swiglu \
  --untie-embeddings-and-output-weights \
  --no-position-embedding \
  --disable-bias-linear \
  --vocab-size 262144 \
  --bf16 \
  --num-experts 128 \
  --moe-router-topk 8 \
  --moe-token-dispatcher-type alltoall \
  --moe-router-load-balancing-type aux_loss \
  --moe-aux-loss-coeff 0.01 \
  --expert-model-parallel-size 4 \
  --tensor-model-parallel-size 1 \
  --pipeline-model-parallel-size 1 \
  --use-distributed-optimizer \
  --overlap-grad-reduce \
  --no-gradient-accumulation-fusion \
  --micro-batch-size 1 \
  --global-batch-size 128 \
  --train-iters 1000 \
  --lr 1e-5 \
  --lr-warmup-iters 50 \
  --lr-decay-style cosine \
  --min-lr 1e-6 \
  --weight-decay 0.01 \
  --clip-grad 1.0 \
  --adam-beta1 0.9 \
  --adam-beta2 0.95 \
  --recompute-granularity full \
  --recompute-method uniform \
  --recompute-num-layers 1 \
  --log-interval 1 \
  --eval-interval 999999 \
  --save-interval 999999 \
  --no-save-optim \
  --no-save-rng \
  --distributed-timeout-minutes 120 \
  --tokenizer-type NullTokenizer \
  --tokenizer-model None
```

For the control case (completes 1000 steps), replace `--expert-model-parallel-size 4 --tensor-model-parallel-size 1` with `--expert-model-parallel-size 1 --tensor-model-parallel-size 4 --sequence-parallel`.

Unfortunately I don't have additional p5e availability for more of these experiments, but I hope this provides enough info to root cause further. Happy to share any other experimental setup info you may need.

**Comment by YangFei1990** [comment]
Thanks for sharing the repo. I will try to reproduce on my end.

In terms of the flex backend, I'm pretty sure they do not use NCCL. Only DeepEPv2 is built on NCCL GIN, but it is not yet available in mcore. Do you know if they could run on AWS EFA? I was aware that there is support issue on EFA for DeepEP early this year but not sure the status now.

**Comment by Aaditya-Singh** [comment]
Thanks for the info and looking into it, for the AWS EFA setup I built DeepEP 2.1.0 from source against NCCL 2.30.7 with [aws-ofi-nccl](https://github.com/aws/aws-ofi-nccl) as the net plugin. Training ran for 852 steps at 3.29 s/step with `--moe_token_dispatcher_type flex` before deadlocking.

I'm not sure if it was really going through DeepEP's transport or silently falling back to alltoall. Either way, the deadlock step and hang signature were identical across all three dispatch backbends (alltoall, deepep, hybridep) so the issue seems to be upstream of dispatch, maybe in mcore's collective ordering on the DP communicator where EP groups are present.
